sled-agent/src/services.rs

```diff
@@ -2932,11 +2929,33 @@ impl ServiceManager {
             //
             // This is (currently) intentional, as the zone filesystem should
             // be destroyed between reboots.
-            let mut rng = rand::thread_rng();
-            let root = all_u2_roots
-                .choose(&mut rng)
-                .ok_or_else(|| Error::U2NotFound)?
-                .clone();
+            let root = if let Some(dataset) = zone.dataset_name() {
+                // If the zone happens to already manage a dataset, then
+                // we co-locate the zone dataset on the same zpool.
+                //
+                // This slightly reduces the underlying fault domain for the
+                // service.
+                let data_pool = dataset.pool();
+                if !all_u2_pools.contains(&data_pool) {
+                    warn!(
+                        log,
+                        "zone dataset requested on a zpool which doesn't exist";
+                        "zone" => &name,
+                        "zpool" => %data_pool
+                    );
+                    return Err(Error::MissingDevice {
```

Comment:
Is there ever a case where we'd want to crate the zone filesystem even if the zpool dataset wasn't found? I can't think of one and believe this makes sense, just putting the question out there for completeness.

Reply:
To be super pedantic:

> even if the zpool dataset wasn't found

This logic is not looking for a dataset, it's looking for a *zpool*. If the zpool doesn't exist, we should not expect to be able to provision any new datasets on it, for any reason.

The sled agent auto-formats new disks to have the necessary zpools for further allocations **before** those disks are made available to the control plane at large, so it should be necessary for us to process these pools before we could make this kinda allocation.

The only situation where I can imagine this causing a "new" error would be:

1. The sled agent boots, formats zpools, informs Nexus about them
2. The sled reboots, we haven't locally parsed which disks are up yet
3. Nexus tries sending a request to provision a new service to that sled
4. This logic runs, and we'd report "the U.2 you're asking for doesn't appear to exist"
5. If Nexus re-sent the request, it should work later, once the sled agent notices that the zpool exists

I *think* the value of the "early error in cases where the disk was truly removed" is more useful, but I could be persuaded to omit this error case, if you think it's a regression?

Reply:
Thanks for the details and pedantry. Always appreciated. :) I think the error is good to keep and in that case nexus should retry anyway I believe, whether with this sled or another. Eventually Inventory will handle the availability of disks and the planner will correct for this scenario I believe